patches/minecraft/net/minecraft/entity/passive/EntityVillager.java.patch

```diff
@@ -17,23 +17,189 @@
++        {
++            professionId = profession.getId();
++        }
++        else if(!professionName.isEmpty() && !oldProfession.equals(professionName))
++        {
++            professionId = 0;
++        }
++        this.func_70938_b(professionId);
+         this.field_70956_bz = p_70037_1_.func_74762_e("Riches");
+         this.field_175563_bv = p_70037_1_.func_74762_e("Career");
+         this.field_175562_bw = p_70037_1_.func_74762_e("CareerLevel");
+         this.field_175565_bs = p_70037_1_.func_74767_n("Willing");
+-
+-        if (p_70037_1_.func_150297_b("Offers", 10))
++        
++        String newProfessionName = net.minecraftforge.fml.common.registry.VillagerRegistry.getProfessionForId(func_70946_n()).getName();
++        if(professionId != func_70946_n()||!(professionName.equals(newProfessionName)||professionName.isEmpty()))
+         {
++            this.field_175563_bv = 0;
++            this.field_175562_bw = 0;
++            this.field_70963_i = null;
++            this.func_70938_b(func_70946_n());
++        }
++        else if (p_70037_1_.func_150297_b("Offers", 10))
++        {
+             NBTTagCompound nbttagcompound1 = p_70037_1_.func_74775_l("Offers");
+             this.field_70963_i = new MerchantRecipeList(nbttagcompound1);
+         }
+@@ -333,14 +356,27 @@
  
-     private void func_175554_cu()
+     public void func_70938_b(int p_70938_1_)
+     {
++        this.field_175563_bv = 0;
++        this.field_175562_bw = 0;
++        this.field_70963_i = null;
+         this.field_70180_af.func_75692_b(16, Integer.valueOf(p_70938_1_));
+     }
+ 
+     public int func_70946_n()
+     {
+-        return Math.max(this.field_70180_af.func_75679_c(16) % 5, 0);
++        return Math.max(this.field_70180_af.func_75679_c(16) % net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession.nextId, 0);
+     }
+ 
++    public void setCareer(int careerId)
++    {
++        this.field_175563_bv = careerId;
++    }
++
++    public int getCareer()
++    {
++        return this.field_175563_bv;
++    }
++
+     public boolean func_70941_o()
      {
-+        //TODO: Hook into VillagerRegistry
-         EntityVillager.ITradeList[][][] aitradelist = field_175561_bA[this.func_70946_n()];
+         return this.field_70952_f;
+@@ -540,15 +576,14 @@
  
+     private void func_175554_cu()
+     {
+-        EntityVillager.ITradeList[][][] aitradelist = field_175561_bA[this.func_70946_n()];
+-
          if (this.field_175563_bv != 0 && this.field_175562_bw != 0)
-@@ -648,6 +650,7 @@
-                     }
+         {
+             ++this.field_175562_bw;
+         }
+         else
+         {
+-            this.field_175563_bv = this.field_70146_Z.nextInt(aitradelist.length) + 1;
++            int careerCount = net.minecraftforge.fml.common.registry.VillagerRegistry.getNumberOfCareers(this.func_70946_n());
++            this.field_175563_bv = this.field_70146_Z.nextInt(careerCount) + 1;
```

Comment:
Any resons why you go the route of using intermediate variables here? makes the patch file harder to read in my opinion.